File: VB/EFCore/ASP.NETCore/Blazor/RuntimeDbChooser.Blazor.Server/Controllers/CustomLogonController.vb
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.ExpressApp.Blazor
Imports DevExpress.ExpressApp.Blazor.SystemModule
Imports DevExpress.ExpressApp.Core
Imports DevExpress.ExpressApp.Security
Imports Microsoft.Extensions.DependencyInjection
Imports RuntimeDbChooser.Services

Namespace RuntimeDbChooser.Blazor.Server.Controllers

    Public Class CustomLogonController
        Inherits BlazorLogonController

        Protected Overrides Sub Accept(ByVal args As SimpleActionExecuteEventArgs)
            Dim targetDataBaseName As String? = CType(Application, BlazorApplication).ServiceProvider.GetRequiredService(Of ILogonParameterProvider)().GetLogonParameters(Of RuntimeDbChooser.Services.IDatabaseNameParameter)().DatabaseName?.Name
            If String.IsNullOrEmpty(targetDataBaseName) Then
                Throw New UserFriendlyException("The DatabaseName is not specified!")
            End If

            CType(Application, BlazorApplication).ServiceProvider.GetRequiredService(Of IObjectSpaceProviderContainer)().Clear()
            MyBase.Accept(args)
        End Sub
    End Class
End Namespace

File: CS/XPO/ASP.NETCore/Blazor/RuntimeDbChooser.Blazor.Server/Controllers/CustomLogonController.cs
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.ExpressApp.Blazor;
using DevExpress.ExpressApp.Blazor.SystemModule;
using DevExpress.ExpressApp.Core;
using DevExpress.ExpressApp.Security;
using Microsoft.Extensions.DependencyInjection;
using RuntimeDbChooser.Services;

namespace RuntimeDbChooser.Blazor.Server.Controllers;
public class CustomLogonController : BlazorLogonController {
    protected override void Accept(SimpleActionExecuteEventArgs args) {
        string? targetDataBaseName = ((BlazorApplication)Application).ServiceProvider.GetRequiredService<ILogonParameterProvider>().GetLogonParameters<IDatabaseNameParameter>().DatabaseName?.Name;
        if(string.IsNullOrEmpty(targetDataBaseName)) {
            throw new UserFriendlyException("The DatabaseName is not specified!");
        }
        ((BlazorApplication)Application).ServiceProvider.GetRequiredService<IObjectSpaceProviderContainer>().Clear();
        base.Accept(args);
    }
}

